scripts/memory_logger.py

```diff
@@ -1,110 +1,149 @@
 #!/usr/bin/env python
 from tempfile import TemporaryFile, SpooledTemporaryFile
-import os, sys, re, socket, time, pickle, csv, uuid, subprocess, argparse, decimal, select, platform
+import os, sys, re, socket, time, pickle, csv, uuid, subprocess, argparse, decimal, select, platform, signal
 
-class LLDB:
-  def __init__(self):
+
+class Debugger:
+  """
+The Debugger class is the entry point to our stack tracing capabilities.
+It determins which debugger to inherit based on parsed arguments and
+platform specs.
+  """
+  def __init__(self, arguments):
+    if arguments.debugger[0] == 'lldb':
+      self.debugger = lldbAPI(arguments)
+    else:
+      self.debugger = DebugInterpreter(arguments)
+
+  def getProcess(self, pid):
+    return self.debugger.getProcess(pid)
+
+  def getStackTrace(self, getProcess_tuple):
+    return self.debugger.getStackTrace(getProcess_tuple)
+
+
+class lldbAPI:
+  def __init__(self, arguments):
     self.debugger = lldb.SBDebugger.Create()
-    self.command_interpreter = self.debugger.GetCommandInterpreter()
-    self.target = self.debugger.CreateTargetWithFileAndArch(None, None)
-    self.listener = lldb.SBListener("event_listener")
-    self.error = lldb.SBError()
+    self.debugger.SetAsync(True)
 
   def __del__(self):
     lldb.SBDebugger.Destroy(self.debugger)
 
+  def getProcess(self, pid):
+    # create and attach to the pid and return our debugger as tuple
+    target = self.debugger.CreateTargetWithFileAndArch(None, None)
+    return target, pid
+
+  def getStackTrace(self, process_tuple):
+    target, pid = process_tuple
+    lldb_results = []
+
+    # reuse the process object if available
+    if target.process.id is not 0:
+      process = target.Attach(lldb.SBAttachInfo(target.process.id), lldb.SBError())
+    else:
+      process = target.Attach(lldb.SBAttachInfo(int(pid)), lldb.SBError())
+
+    # test if we succeeded at attaching to PID process
+    if process:
+      # grab thread information
+      lldb_results.append(process.GetThreadAtIndex(0).__str__())
+      # iterate through all frames and collect back trace information
+      for i in xrange(process.GetThreadAtIndex(0).GetNumFrames()):
+        lldb_results.append(process.GetThreadAtIndex(0).GetFrameAtIndex(i).__str__())
+      # unfortunately we must detach each time we perform a stack trace
+      # this is a bug in LLDB's Python API.
+      process.Detach()
+      return '\n'.join(lldb_results)
+    else:
+      return ''
+
+class DebugInterpreter:
+  """
+Currently, interfacing with LLDB via subprocess is impossible. This is due to lldb not printing
+to stdout, or stderr when displaying the prompt to the user (informing the user, the debugger
+is ready to receive input). However, this class may someday be able to, which is why
+the self.debugger variable is present.
+  """
```

Comment:
Thanks for the comment, you'll be happy you put this here in 6 months when this breaks again.